[59468] Danger Confirmation Dialog implementation

myabc opened the pull request:
Based on `FeedbackDialog`. Wraps `Primer::Alpha::Dialog`, providing additional standardised behaviour for scenarios where a user performs a "dangerous" action.

Includes initial Lookbook previews and component tests.

Implements `DangerConfirmationDialogFormHelperElement` Catalyst controller providing checkbox confirmation behaviour.

### What are you trying to accomplish?
<!-- Provide a description of the changes. -->

### Screenshots


https://raw.githubusercontent.com/opf/primer_view_components/d232bfc77ee23bcde7e00ac3bd50c70bc2f4d034/.playwright/screenshots/snapshots.test.ts-snapshots/primer/open_project/danger_confirmation_dialog/default/light.png

### Integration
<!-- Does this change require any updates to code in production? -->

#### List the issues that this change affects.

https://community.openproject.org/projects/openproject/work_packages/58680

#### Risk Assessment
  <!-- Please select from one of the following and detail why this level was chosen -->

- [x] **Low risk** the change is small, highly observable, and easily rolled back.
- [ ] **Medium risk** changes that are isolated, reduced in scope or could impact few users. The change will not impact library availability.
- [ ] **High risk** changes are those that could impact customers and SLOs, low or no test coverage, low observability, or slow to rollback.

### What approach did you choose and why?
<!-- This section is a place for you to describe your thought process in making these changes.
     List any tradeoffs you made to take on or pay down tech debt.
     Identify any work you did to mitigate risk.
     Describe any alternative approaches you considered and why you discarded them. -->

### Anything you want to highlight for special attention from reviewers?
<!-- This is your chance to identify remaining risks and confess any uncertainties you may have about the correctness of the changes.
     Highlight anything on which you would like a second (or third) opinion.
     Keep in mind how many component uses cases may be affected by your changes when assessing risk. -->

### Accessibility
<!--
  You may remove this section and the "Accessibility" heading above _only_ if the changes in this pull request do not impact UI. Delete all those that don't apply.
  If there are any accessibility-related updates, please describe them here.
-->
- **Fixes axe scan violation** - This change fixes an existing [axe scan](https://thehub.github.com/epd/engineering/dev-practicals/frontend/accessibility/readiness-routine/development/#axe-scans) violation.
- **No new axe scan violation** - This change does not introduce any new [axe scan](https://thehub.github.com/epd/engineering/dev-practicals/frontend/accessibility/readiness-routine/development/#axe-scans) violations.
- **New axe violation** - This change introduces a new [axe scan](https://thehub.github.com/epd/engineering/dev-practicals/frontend/accessibility/readiness-routine/development/#axe-scans) violation. Please describe why the violation cannot be resolved below.

### Merge checklist

- [x] Added/updated tests
  - [X] Component tests
  - [x] System tests
- [x] Added/updated documentation
- [x] Added/updated previews (Lookbook)
- [x] Tested in Chrome
- [x] Tested in Firefox
- [x] Tested in Safari
- [x] Tested in Edge

**Additional TODOs/Questions for Discussion**

- [x] Determine how to handle I18n _Resolution: leave to caller_
- [X] ~~Test with Rails form builders (?)~~ _Resolution: test once upstream_

HDinger commented on previews/primer/open_project/danger_confirmation_dialog_preview.rb:
```diff
@@ -0,0 +1,68 @@
+      # @label Playground
+      # @param icon [Symbol] octicon
+      # @param icon_color [Symbol] select [default, muted, subtle, accent, success, attention, severe, danger, open, closed, done, sponsors, on_emphasis, inherit]
+      # @param loading_state [Boolean] toggle
```
I think we should not show the loading_state as option as it does not make much sense for the danger dialog

HDinger commented on previews/primer/open_project/danger_confirmation_dialog_preview/playground.html.erb:
```diff
@@ -0,0 +1,12 @@
+  <% end %>
+  <% if show_additional_content %>
+    <% dialog.with_additional_content(display: :inline) do %>
+      <%= render(Primer::Alpha::Banner.new) { "Some additional content below" } %>
```
This example looks somehow weird:

<img width="512" alt="Bildschirmfoto 2024-12-12 um 14 14 44" src="https://github.com/user-attachments/assets/07d0dc2c-c404-4030-b2e1-7d84452bc24c" />

myabc replied:
I've updated the Preview example - I think we can leave it up to the caller not to override styling of additional details with `:inline`.

<img width="508" alt="image" src="https://github.com/user-attachments/assets/630ecfb8-6a23-4f67-839d-8ead08b2c0d5" />

HDinger reviewed (changes requested):
Hi @myabc 

nice job, it looks really good 👍 I only have some minor remarks.

myabc commented on app/components/primer/open_project/danger_confirmation_dialog.rb:
```diff
@@ -0,0 +1,83 @@
+          system_arguments[:classes],
+          "DangerConfirmationDialog"
+        )
+        @system_arguments[:id] = @dialog_id
```
I've moved the assignment. Let me know if that works for you?

myabc commented:
@HDinger I've addressed most of the issues you raised.

In addition to the couple of questions inline, a couple general questions:

* Should we use `checkbox`/`Checkbox` or `check_box`/`CheckBox`? The HTML spec uses `checkbox`, and is what I would normally prefer. It's one less keystroke for our end users. However the latter would make us consistent with the other Primer components.
* `DangerConfirmationDialog` is a pretty long, cumbersome name IMO. "confirmation" is probably redundant, given that our slots also our named `confirmation_message` and `confirmation_dialog`. Could we consider going with just (an opinionated) `DangerDialog`?

myabc commented on app/components/primer/open_project/danger_confirmation_dialog.html.erb:
```diff
@@ -0,0 +1,19 @@
+      <scrollable-region data-labelled-by="<%= labelledby %>">
+        <%= render(Primer::Alpha::Dialog::Body.new) do %>
+          <%= confirmation_message %>
+          <% if additional_details? %>
```
I will rename the `FeedbackDialog` `additional_content` slot to `additional_details` in a separate PR.

myabc commented on app/components/primer/open_project/danger_confirmation_dialog/confirmation_checkbox.rb:
```diff
@@ -0,0 +1,37 @@
+
+        def call
+          render(Primer::BaseComponent.new(**@system_arguments)) do
+            render(Primer::Alpha::CheckBox.new(**@checkbox_arguments.merge(label: trimmed_content)))
```
@HDinger As I've implemented it, we build the checkbox label based on (`trimmed_`)`content` - this allows for a similar API to the other slots, i.e.:

```ruby
      dialog.with_show_button { "Click me" }
      dialog.with_confirmation_checkbox { "I confirm this deletion" }
```

An alternative would be to add a `label_text` string parameter - either on `DangerConfirmationDialog` itself - or on the slot. I am not sure "what feels" more idiomatic?

```ruby
  DangerConfirmationDialog.new(label_text: "I confirm this deletion")
  dialog.with_confirmation_checkbox(label_text: "I confirm this deletion")
```

myabc replied:
Also, do we want to define a `I18n.t` default for the label? Or is it fine to always leave it to the caller?

HDinger replied:
Currently you can write something like `dialog.with_confirmation_checkbox` which will render a checkbox without any label. So we either have to provide a default or enforce that the user provides a string. Since the default is rather special, we probably have to enforce the the user to pass an argument for now. 

With regards to the syntax, I like the `dialog.with_confirmation_checkbox { "I confirm this deletion" }` notation, but if it is easier to enfoce a text with a named parameter, I am also fine with that.

HDinger commented on .changeset/brave-eggs-obey.md:
```diff
@@ -0,0 +1,6 @@
+---
+"@openproject/primer-view-components": patch
```
please make that a `minor`  release

HDinger commented on test/components/component_test.rb:
```diff
@@ -8,6 +8,13 @@ class PrimerComponentTest < Minitest::Test
+      end
+      component.with_confirmation_checkbox { "Really do this?" }
+    }],
+    [Primer::OpenProject::DangerConfirmationDialog::ConfirmationCheckbox, {}],
```
I think we can ignore this from the tests. At the bottom of the file, there is a list of exlcuded classes, where you can simply add this.

myabc replied:
I went ahead and ignored the internal `Primer::OpenProject::DangerConfirmationDialog::FormWrapper` component, as you suggested. I've am not sure why that is preferable though?

HDinger replied:
We usually ignore those components that are more helper components of their parents and don't make too much sense standalone. Those will be tested in the scope of their parents and not individually. I think it is fine the way you did it 👍

HDinger commented on app/components/primer/open_project/danger_confirmation_dialog.html.erb:
```diff
@@ -0,0 +1,19 @@
+      </scrollable-region>
+      <%= render(Primer::Alpha::Dialog::Footer.new(show_divider: false)) do %>
+        <%= render(Primer::Beta::Button.new(data: { "close-dialog-id": dialog_id })) { I18n.t("button_close") } %>
+        <%= render(Primer::Beta::Button.new(scheme: :danger, data: { "submit-dialog-id": dialog_id })) { I18n.t("button_delete_permanently") } %>
```
The button currently does not submit the form when being pressed. `"submit-dialog-id"` only closes the Dialog but does not trigger any form update.

myabc replied:
Fixed in [aa84913](https://github.com/opf/primer_view_components/pull/214/commits/aa8491308535550bedfdde88eec2f102f3f7d074)

HDinger reviewed (changes requested):
I left some remarks and will check what is happening with the snapshots.

HDinger reviewed (approved):
Nice job 🌟  I really like the way it can be used 👍
